File: packages/agent/agent/tools/prepare_batch_registration.ts
import { randomUUID } from "node:crypto";

import { getDb } from "@clara-financas/db";
import { financialActionProposals } from "@clara-financas/db/schema/financial-action";
import { batches, documents, transactions } from "@clara-financas/db/schema/ledger";
import { forTenant } from "@clara-financas/db/tenant-scope";
import {
  formatCents,
  formatDocumentLabel,
  type ChecksumReport,
  type StatementBalanceReport,
} from "@clara-financas/ledger";
import { and, count, eq } from "drizzle-orm";
import { defineTool } from "eve/tools";
import { z } from "zod";

import { notFound, refused } from "../lib/errors";
import { requireTenantCaller } from "../lib/tenant";

export default defineTool({
  description:
    "Prepares the canonical registration proposal for one draft financial document. Call immediately before commit_batch, then pass the returned proposalId to commit_batch. This freezes the document revision, verification and number of entries shown in the approval.",
  inputSchema: z.object({ batchId: z.string().min(1) }),
  async execute(input, ctx) {
    const { tenantId, userId } = requireTenantCaller(ctx);
    return forTenant(
      tenantId,
      async (tx) => {
        const [found] = await tx
          .select({
            id: batches.id,
            status: batches.status,
            updatedAt: batches.updatedAt,
            checksumReport: batches.checksumReport,
            issuer: documents.issuer,
            documentKind: documents.kind,
            periodEnd: batches.periodEnd,
            dueDate: batches.dueDate,
            openingBalance: batches.openingBalance,
            closingBalance: batches.closingBalance,
          })
          .from(batches)
          .innerJoin(documents, eq(documents.id, batches.documentId))
          .where(and(eq(batches.id, input.batchId), eq(batches.tenantId, tenantId)))
          .limit(1);
        if (found === undefined) {
          return notFound("lote_nao_encontrado", `Nenhum documento com o id ${input.batchId}.`);
        }
        if (found.status !== "proposed") {
          return refused("lote_ja_decidido", "Este documento já foi decidido.");
        }

        const countRows = await tx
          .select({ value: count() })
          .from(transactions)
          .where(
            and(
              eq(transactions.batchId, found.id),
              eq(transactions.tenantId, tenantId),
              eq(transactions.status, "proposed"),
            ),
          );
        const transactionCount = countRows[0]?.value ?? 0;
        const verification = found.checksumReport as (ChecksumReport | StatementBalanceReport) | null;
        const statementBalance =
          verification !== null && "kind" in verification && verification.kind === "statement_balance"
            ? verification
            : null;
        const checksum = statementBalance === null ? (verification as ChecksumReport | null) : null;
        const proposalId = `act_${randomUUID().replace(/-/g, "").slice(0, 24)}`;
        const expiresAt = new Date(Date.now() + 30 * 60_000);
        const invoiceLabel = formatDocumentLabel(found);
        const payload = {
          transactionCount,
          checksumResult:
            statementBalance?.result === "match"
              ? "match"
              : statementBalance?.result === "mismatch"
                ? "mismatch"
                : checksum?.result ?? null,
          declaredTotalCents: checksum?.declaredTotal ?? null,
          extractedTotalCents: checksum?.extractedTotal ?? null,
          differenceCents: checksum?.difference ?? null,
          issuer: found.issuer,
          documentKind: found.documentKind,
          invoiceLabel,
          openingBalanceCents: found.openingBalance,
          closingBalanceCents: found.closingBalance,
          statementBalance,
        };
        await tx.insert(financialActionProposals).values({
          id: proposalId,
          tenantId,
          batchId: found.id,
          operation: "register_invoice",
          entityRevision: found.updatedAt,
          payload,
          preparedBy: `human:${userId}`,
          expiresAt,
        });

        return {
          proposalId,
          actionId: proposalId,
          operation: "register_invoice" as const,
          batchId: found.id,
          entityRevision: found.updatedAt.toISOString(),
          expiresAt: expiresAt.toISOString(),
          issuer: found.issuer,
          documentKind: found.documentKind,
          invoiceLabel,
          openingBalanceCents: found.openingBalance,
          closingBalanceCents: found.closingBalance,
          transactionCount,
          checksumResult:
            statementBalance?.result === "match"
              ? "match"
              : statementBalance?.result === "mismatch"
                ? "mismatch"
                : checksum?.result ?? null,
          declaredTotalCents: checksum?.declaredTotal ?? null,
          extractedTotalCents: checksum?.extractedTotal ?? null,
          extractedTotalFormatted:
            checksum === null ? null : formatCents(checksum.extractedTotal),
          differenceCents: checksum?.difference ?? null,
          ...(statementBalance === null ? {} : { statementBalance }),
          next: "Call commit_batch with this proposalId. Do not ask for a prose confirmation.",
        };
      },
      getDb(),
    );
  },
});

File: services/agent/clara/tools/prepare_batch_registration.py
"""Prepara a proposta canônica de registro — porta de `prepare_batch_registration.ts`.

Chamada imediatamente antes de `commit_batch`: congela a revisão do
documento, a conferência e a contagem de lançamentos mostrados na aprovação.
"""

from __future__ import annotations

import uuid
from dataclasses import dataclass
from datetime import UTC, datetime, timedelta

from sqlalchemy import func, select
from sqlalchemy.orm import Session

from clara.db.models import Batch, Document, FinancialActionProposal, Transaction
from clara.ledger.invoice_label import format_document_label
from clara.ledger.money import format_cents
from clara.tools.errors import ToolError, not_found, refused

PROPOSAL_TTL = timedelta(minutes=30)


@dataclass
class PreparedRegistration:
    proposal_id: str
    batch_id: str
    entity_revision: datetime
    expires_at: datetime
    issuer: str | None
    document_kind: str
    invoice_label: str
    transaction_count: int
    checksum_result: str | None
    declared_total_cents: int | None
    extracted_total_cents: int | None
    extracted_total_formatted: str | None
    difference_cents: int | None


def prepare_batch_registration(
    session: Session, tenant_id: str, user_id: str, batch_id: str
) -> PreparedRegistration | ToolError:
    row = session.execute(
        select(Batch, Document.issuer, Document.kind)
        .join(Document, Document.id == Batch.document_id)
        .where(Batch.id == batch_id, Batch.tenant_id == tenant_id)
    ).first()

    if row is None:
        return not_found("lote_nao_encontrado", f"Nenhum documento com o id {batch_id}.")

    batch, issuer, document_kind = row
    if batch.status != "proposed":
        return refused("lote_ja_decidido", "Este documento já foi decidido.")

    transaction_count = session.execute(
        select(func.count(Transaction.id)).where(
            Transaction.batch_id == batch.id,
            Transaction.tenant_id == tenant_id,
            Transaction.status == "proposed",
        )
    ).scalar_one()

    report = batch.checksum_report or {}
    is_statement = report.get("kind") == "statement_balance"
    checksum = None if is_statement else report or None

    proposal_id = f"act_{uuid.uuid4().hex[:24]}"
    expires_at = datetime.now(UTC) + PROPOSAL_TTL
    invoice_label = format_document_label(
        issuer=issuer, due_date=batch.due_date, period_end=batch.period_end,
        document_kind=document_kind,
    )

    checksum_result = report.get("result") if report else None
    declared_total = checksum.get("declared_total") if checksum else None
    extracted_total = checksum.get("extracted_total") if checksum else None
    difference = checksum.get("difference") if checksum else None

    payload = {
        "transaction_count": transaction_count,
        "checksum_result": checksum_result,
        "declared_total_cents": declared_total,
        "extracted_total_cents": extracted_total,
        "difference_cents": difference,
        "issuer": issuer,
        "document_kind": document_kind,
        "invoice_label": invoice_label,
    }

    session.add(
        FinancialActionProposal(
            id=proposal_id, tenant_id=tenant_id, batch_id=batch.id,
            operation="register_invoice", entity_revision=batch.updated_at,
            payload=payload, prepared_by=f"human:{user_id}", expires_at=expires_at,
        )
    )
    session.flush()

    return PreparedRegistration(
        proposal_id=proposal_id, batch_id=batch.id, entity_revision=batch.updated_at,
        expires_at=expires_at, issuer=issuer, document_kind=document_kind,
        invoice_label=invoice_label, transaction_count=transaction_count,
        checksum_result=checksum_result, declared_total_cents=declared_total,
        extracted_total_cents=extracted_total,
        extracted_total_formatted=(
            None if extracted_total is None else format_cents(extracted_total)
        ),
        difference_cents=difference,
    )
